### PR TITLE
New: Date Picker Component

### PR DESCRIPTION
**Pull request by davepoon**
- Adding a new DatePicker Component.
- This component is reusing the third party React DatePicker component lib.
- This is not 'reduxified' yet.
- Custom style overrides will be added.

### CODE REVIEW

**Comment by camjc** [comment]
Got some conflicts mate, can you please `git fetch` then `git rebase -i origin/master` and resolve any merge conflicts then `git push -f` your one commit. @davepoon

**Comment by camjc** [comment]
Need to fix up your commits (should be 1 commit),
write tests for your example so coverage is still 100%,
add style overrides.

LGTM so far, nice stuff.

File: src/components/Main.js [hunk 1]
```diff
@@ -53,9 +57,22 @@ class AppComponent extends React.Component {
...
+  }
+
+  setSelectedFlavours(newValue) {
+    this.setState({ selectedFlavours: newValue });
+  }
+
+  setSelectedDate(newValue) {
+    this.setState({ startDate: newValue });
+  }
+
```
Comment (omgaz) [inline src/components/Main.js]: I think you've had a bad rebase?  :P

File: src/components/Main.js [hunk 2]
```diff
@@ -53,9 +57,22 @@ class AppComponent extends React.Component {
...
     };
   }
 
+  setSelectedCountry(newValue) {
+    this.setState({ selectedCountry: newValue.value });
+  }
+
+  setSelectedFlavours(newValue) {
+    this.setState({ selectedFlavours: newValue });
+  }
```
Comment (omgaz) [inline src/components/Main.js]: I think you've had a bad rebase? :P

File: src/components/distributionEntry.js [hunk 3]
```diff
@@ -4,13 +4,15 @@ require('styles/_bootstrap-custom.scss');
...
 require('styles/_react-toggle-custom.scss');
+require('styles/_react-datepicker-custom.scss');
 
 import Button from 'react-bootstrap/lib/Button';
 import Checkbox from 'react-icheck/lib/Checkbox';
 import ConfirmModal from 'components/adslotUi/ConfirmModalComponent';
 import FormGroup from 'components/adslotUi/FormGroupComponent';
 import ListPicker from 'components/adslotUi/ListPickerComponent';
 import ListPickerPure from 'components/adslotUi/ListPickerPureComponent';
+import DatePicker from 'react-datepicker/dist/react-datepicker';
```
Comment (omgaz) [inline src/components/distributionEntry.js]: Can we keep this list alphabetical?

File: src/components/Main.js (hunk 2 above)
Comment (davepoon) [inline src/components/Main.js]: I gonna make it into one commit, thanks for the feedback

File: src/styles/_react-datepicker-custom.scss [hunk 4]
```diff
@@ -0,0 +1,55 @@
+@import './variable';
+@import '../../node_modules/react-datepicker/dist/react-datepicker.css';
+
+$datepicker-input-border-color: $color-border-lighter;
+$datepicker-input-border-radius: $border-radius-base;
+$datepicker-input-border-focus:  $color-border;
+$datepicker-input-height: 26px;
+$datepicker-padding-vertical: 4px;
+$datepicker-padding-horizontal: 7px;
```
Comment (omgaz) [inline src/styles/_react-datepicker-custom.scss]: Can we not use these:

```
@variable.scss
$padding-base-vertical: 4px;
$padding-base-horizontal: 8px;
```

File: src/components/distributionEntry.js (hunk 3 above)
Comment (davepoon) [inline src/components/distributionEntry.js]: Sure.

File: src/styles/_react-datepicker-custom.scss [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+$datepicker-text-color: $color-text-form;
+$datepicker-font-weight: $font-weight-medium;
+$datepicker-font-size: $font-size-base;
+$datepicker-header-background-color: $color-gray-lightest;
```
Comment (omgaz) [inline src/styles/_react-datepicker-custom.scss]: `$color-background`

File: src/styles/_react-datepicker-custom.scss [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  font-size: $datepicker-font-size;
+  font-weight: $datepicker-font-weight;
+}
+
+.datepicker__header {
+  background-color: $datepicker-header-background-color;
+}
+
+.datepicker__day--outside-month {
+  color: $color-gray;
```
Comment (omgaz) [inline src/styles/_react-datepicker-custom.scss]: Don't use the colours directly, use the informative ones e.g. $color-disabled or something that has meaning to what you're trying to convey.

File: src/styles/_react-datepicker-custom.scss [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+.datepicker__day--selected {
+  background-color: $color-primary;
+  border-radius: $datepicker-input-border-radius;
+
+  &:hover {
+    border-radius: $datepicker-input-border-radius;
+  }
+}
```
Comment (omgaz) [inline src/styles/_react-datepicker-custom.scss]: You can save repetition with a lonely ampersand:

```
.datepicker__day--selected {
  background-color: $color-primary;

  &,
  &:hover {
    border-radius: $datepicker-input-border-radius;
  }
}
```

File: src/styles/_react-datepicker-custom.scss [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  color: $datepicker-text-color;
+  font-family: $font-family-sans-serif;
+  font-size: $datepicker-font-size;
+  font-weight: $datepicker-font-weight;
+
+  &:focus {
+    border-color: $datepicker-input-border-focus;
+    box-shadow: none;
+  }
+}
```
Comment (omgaz) [inline src/styles/_react-datepicker-custom.scss]: I'd group this file using ampersand like so

```scss
.datepicker {
  &__header {}
  
  __day--{
    &--outside-month {
      &,
      &:hover {}
    }
  }
  
  &--selected {}
  
  &__current-month {}
  
  &__input {}
}
```

File: test/components/MainTest.js [hunk 9]
```diff
@@ -9,6 +9,7 @@ import {
   Modal,
   Radio,
   RadioGroup,
+  DatePicker,
```
Comment (omgaz) [inline test/components/MainTest.js]: Alphabetical again please.

**Comment by omgaz** [comment]
Super-awesome job! Have a :cookie:

**Comment by alextkachuk** [comment]
Looks good for me!

**Comment by davepoon** [comment]
I have updated the Adslot UI datepicker scss due to the css refactor and css class name change in the latest version(v0.25.0) of the third party react datepicker library, which could break my css overrides.
And also,  's' duration unit is added to the property-units rule in sass-lint.yml, so css transition unit is allowed in our components.

**Comment by omgaz** [comment]
Great job, Dave. If you have time in your next PR or otherwise it'd be great to see that missing branch added to keep our coverage at 100% :)

**Comment by davepoon** [comment]
Okay, cool~

**Comment by omgaz** [comment]
Hey, just seen this has been merged already. Has it beed design reviewed by Pedros, is it feature complete?

**Comment by davepoon** [comment]
Oh, no it hasn't been reviewd by Pedros, I thought it's been reviewed... sorry about that I am new to the review process...
It is fine to revert the merge in the master branch, and reopen this PR to get Pedros review,
and I can also add the codecov piece and add bootstrap .form-control to unify the datepicker design.
And get another round of review.

**Pull request merged**